### PR TITLE
Add `position` API for VWRS to allow interaction with native apps

### PR DESCRIPTION
**Pull request by todor-macrometa**
Also added the GET /status API, which already a part of `photoniq-vwrs/master` and `photoniq-vwrs/release-1.26`, but not `docs`.

### CODE REVIEW

**Review by joubert-macrometa — APPROVED** [review]

File: lib/openapi/revisions/vwrs-spec-1.26.0-20240430.json
Comment (diana-macrometa) [inline lib/openapi/revisions/vwrs-spec-1.26.0-20240430.json:1]: What is this new file for?

File: static/openapi/vwrs-spec.json [hunk 1]
```diff
@@ -467,6 +467,70 @@
...
           }
         }
+      },
+      "GetStatusResponse": {
+        "type": "object",
+        "properties": {
+          "avg_waiting_time": {
+            "type": "number",
+            "format": "float",
+            "description": "The average time users spend in the virtual waiting room before being granted access to the origin service (in milliseconds).",
```
Comment (diana-macrometa) [inline static/openapi/vwrs-spec.json]: @joubert-macrometa, what is the standard for indicating time units? In this edit, we do it two different ways:

- Blah blah blah **(in time unit)**.
- Blah **in time unit** blah blah blah.

File: static/openapi/vwrs-spec.json [hunk 2]
```diff
@@ -467,6 +467,70 @@
...
+            "example": 19871.8
+          },
+          "backoff_interval": {
+            "type": "integer",
+            "description": "Indicates the interval in seconds after which the EdgeWorker should call the status API call to the vwrs server. This is used to reduce server load.",
```
Comment (diana-macrometa) [inline static/openapi/vwrs-spec.json]:
```suggestion
            "description": "Indicates the interval in seconds after which the EdgeWorker sends the status API call to the VWRs server. This is used to reduce server load.",
```

File: static/openapi/vwrs-spec.json [hunk 3]
```diff
@@ -467,6 +467,70 @@
...
+            "example": 1
+          },
+          "position": {
+            "type": "integer",
+            "description": "The client's current position in the waiting room.",
```
Comment (diana-macrometa) [inline static/openapi/vwrs-spec.json]: Are we calling them users or clients? I think we also use visitors somewhere. In this PR, we're using at least _client_ and _users_.

File: static/openapi/vwrs-spec.json [hunk 4]
```diff
@@ -467,6 +467,70 @@
...
+            "example": 99359
+          },
+          "rate_limit": {
+            "type": "integer",
+            "description": "The rate limit for the domain. The maximum users per second allowed to reach the origin.",
+            "example": 5
+          },
+          "waiting_room_interval": {
+            "type": "integer",
+            "description": "The time interval in seconds when the UI reloads itself to get the waiting room status.",
```
Comment (diana-macrometa) [inline static/openapi/vwrs-spec.json]: Whose UI? Our UI? The native UI?

File: static/openapi/vwrs-spec.json [hunk 5]
```diff
@@ -467,6 +467,70 @@
...
+      },
+      "GetAccessResponse": {
+        "type": "object",
+        "required": [
+          "position"
+        ],
+        "properties": {
+          "position": {
+            "type": "number",
+            "description": "Position in the waiting room. 0 means the user doesn't need to wait in waiting room any longer and can go to the origin."
```
Comment (diana-macrometa) [inline static/openapi/vwrs-spec.json]:
```suggestion
            "description": "Position in the waiting room. 0 means the user doesn't need to wait in the waiting room any longer and can go to the origin."
```

Comment (diana-macrometa) [reply static/openapi/vwrs-spec.json]: So, they don't need to wait, but are they automatically advanced? Or is this something to tell the native app to advance them?

File: static/openapi/vwrs-spec.json [hunk 6]
```diff
@@ -467,6 +467,70 @@
...
+            "type": "number",
+            "description": "Number of users currently in the waiting_room."
+          },
+          "visitor_id": {
+            "type": "number",
+            "description": "The user's session ID for this access request."
+          },
+          "refresh_interval": {
+            "type": "number",
+            "description": "Specifies the time interval, in seconds, for subsequent checks of the session status."
```
Comment (diana-macrometa) [inline static/openapi/vwrs-spec.json:531]: And here's a third way we are stating the time unit.

File: static/openapi/vwrs-spec.json [hunk 7]
```diff
@@ -1164,6 +1228,194 @@
...
+            "description": "The position ID in the waiting room."
+          },
+          {
+            "in": "query",
+            "name": "created_at",
+            "required": true,
+            "schema": {
+              "type": "integer"
+            },
+            "description": "The creation timestamp when user first time entered the waiting room."
```
Comment (diana-macrometa) [inline static/openapi/vwrs-spec.json]:
```suggestion
            "description": "The creation timestamp when the user first time entered the waiting room."
```

File: static/openapi/vwrs-spec.json [hunk 8]
```diff
@@ -1164,6 +1228,194 @@
...
+            "description": "The domain_key of the domain."
+          },
+          {
+            "in": "query",
+            "name": "sid",
+            "required": true,
+            "schema": {
+              "type": "integer"
+            },
+            "description": "The position ID in the waiting room."
```
Comment (diana-macrometa) [inline static/openapi/vwrs-spec.json]: You mean the user's position ID? Or something else?

File: static/openapi/vwrs-spec.json [hunk 9]
```diff
@@ -1164,6 +1228,194 @@
...
+          "500": {
+            "description": "Internal server error."
+          }
+        }
+      }
+    },
+    "/api/vwr/v1/position/{domain_key}": {
+      "get": {
+        "summary": "Provide initial position in the waiting room",
+        "description": "Provide initial position in the waiting room.",
```
Comment (diana-macrometa) [inline static/openapi/vwrs-spec.json]: Why "provide" rather than "get," which is used earlier?

File: static/openapi/vwrs-spec.json [hunk 10]
```diff
@@ -1164,6 +1228,194 @@
...
+            "required": true,
+            "schema": {
+              "type": "string"
+            },
+            "description": "The unique key of a domain."
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "Successfully obtained waiting room data.",
```
Comment (diana-macrometa) [inline static/openapi/vwrs-spec.json]:
```suggestion
            "description": "Successfully retrieved waiting room data.",
```

File: static/openapi/vwrs-spec.json [hunk 11]
```diff
@@ -1164,6 +1228,194 @@
...
+          "500": {
+            "description": "Internal server error."
+          }
+        }
+      }
+    },
+    "/api/vwr/v1/position/{domain_key}/{visitor_id}": {
+      "get": {
+        "summary": "Provide current position in the waiting room",
+        "description": "Provides the visitor's position in the waiting room.\\n\\nA position of 0 indicates that the visitor is no longer in the waiting room and has gained access to the requested resource.",
```
Comment (diana-macrometa) [inline static/openapi/vwrs-spec.json]: And here's that third one. Are we talking about clients, users, or visitors? Pick one and stick with it throughout.

File: static/openapi/vwrs-spec.json [hunk 12]
```diff
@@ -1164,6 +1228,194 @@
...
+            "required": true,
+            "schema": {
+              "type": "number"
+            },
+            "description": "The user's session ID."
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "Successfully obtained waiting room data.",
```
Comment (diana-macrometa) [inline static/openapi/vwrs-spec.json]:
```suggestion
            "description": "Successfully retrieved waiting room data.",
```

**Review by diana-macrometa — CHANGES_REQUESTED** [review]

File: static/openapi/vwrs-spec.json (hunk 2 above)
Comment (joubert-macrometa) [reply static/openapi/vwrs-spec.json]: Remove the "This is used to reduce server load."  Does not need to be in document.

Comment (joubert-macrometa) [reply static/openapi/vwrs-spec.json]: Fix interval in seconds to be consistent.

File: static/openapi/vwrs-spec.json (hunk 3 above)
Comment (joubert-macrometa) [inline static/openapi/vwrs-spec.json]:
```suggestion
            "description": "The user's current position in the waiting room.",
```

File: static/openapi/vwrs-spec.json (hunk 4 above)
Comment (joubert-macrometa) [inline static/openapi/vwrs-spec.json]:
```suggestion
            "description": "The time interval in seconds when the waiting room status page is reloaded.",
```

Comment (joubert-macrometa) [reply static/openapi/vwrs-spec.json]: Fix the time interval to be consistent.

File: static/openapi/vwrs-spec.json (hunk 5 above)
Comment (joubert-macrometa) [reply static/openapi/vwrs-spec.json]: This is your (the users) position in the queue. When you enter the queue, you are in the 100s position. A few seconds later, you are in the 70s position, and so on. In the end, VWR has determined that you are out of the queue. This means that when you make the API to determine your position and your position is zero, it means VWR has moved you out of the waiting room.

File: static/openapi/vwrs-spec.json [hunk 13]
```diff
@@ -467,6 +467,70 @@
...
+          "position"
+        ],
+        "properties": {
+          "position": {
+            "type": "number",
+            "description": "Position in the waiting room. 0 means the user doesn't need to wait in waiting room any longer and can go to the origin."
+          },
+          "waiting_time": {
+            "type": "number",
+            "description": "Expected time to spend in waiting room (in seconds)."
```
Comment (joubert-macrometa) [inline static/openapi/vwrs-spec.json]: Consistent time

File: static/openapi/vwrs-spec.json [hunk 14]
```diff
@@ -467,6 +467,70 @@
...
+          },
+          "queue_depth": {
+            "type": "number",
+            "description": "Number of users currently in the waiting_room."
```
Comment (joubert-macrometa) [inline static/openapi/vwrs-spec.json]: Should that be "waiting room"?

File: static/openapi/vwrs-spec.json (hunk 6 above)
Comment (joubert-macrometa) [reply static/openapi/vwrs-spec.json:531]: @todor-macrometa can we change visitior_id to request_id to be consistent with the status API below this one.

File: static/openapi/vwrs-spec.json (hunk 5 above)
Comment (joubert-macrometa) [reply static/openapi/vwrs-spec.json]: Maybe go with:

The user's current position in the waiting room.  0 means the user doesn't need to wait in the waiting room any longer and can go to the origin.

The first sentence is what we used below in the "status" api

File: static/openapi/vwrs-spec.json (hunk 7 above)
Comment (joubert-macrometa) [reply static/openapi/vwrs-spec.json]: Can we include if this is in seconds or milli seconds?  

In other parts of this document, we said:
"The Unix timestamp when ...."

so

"The Unix creation timestamp"???  or "The creation Unix timestamp".  Not sure the second one is right

File: static/openapi/vwrs-spec.json (hunk 9 above)
Comment (joubert-macrometa) [reply static/openapi/vwrs-spec.json]: Yes.  Get.

File: static/openapi/vwrs-spec.json (hunk 11 above)
Comment (joubert-macrometa) [reply static/openapi/vwrs-spec.json]: This should be user's.

Comment (joubert-macrometa) [inline static/openapi/vwrs-spec.json]:
```suggestion
        "description": "Provides the user's position in the waiting room.\\n\\nA position of 0 indicates that the user is no longer in the waiting room and has gained access to the requested resource.",
```

File: lib/openapi/revisions/vwrs-spec-1.26.0-20240430.json
Comment (todor-macrometa) [reply lib/openapi/revisions/vwrs-spec-1.26.0-20240430.json:1]: This guide says (in step 12) that I should also add the file that I've used to generate the vwrs-spec file:
https://www.notion.so/macrometa/Steps-to-update-PhotonIQ-VWRS-API-reference-documentation-e451f63422c64a1d9e29ee12b4f1c05c

File: static/openapi/vwrs-spec.json (hunk 7 above)
Comment (todor-macrometa) [reply static/openapi/vwrs-spec.json]: I'll leave the second one for now. First sounds a bit like "the timestamp when unix was created"

File: static/openapi/vwrs-spec.json [hunk 15]
```diff
@@ -1164,6 +1228,194 @@
...
+            "description": "The user's position ID in the waiting room."
+          },
+          {
+            "in": "query",
+            "name": "created_at",
+            "required": true,
+            "schema": {
+              "type": "integer"
+            },
+            "description": "The creation Unix timestamp, in seconds, when the user first time entered the waiting room."
```
Comment (diana-macrometa) [inline static/openapi/vwrs-spec.json]:
```suggestion
            "description": "Unix timestamp, in seconds, when the user first time entered the waiting room."
```

**Review by diana-macrometa — APPROVED** [review]
One minor edit, then g2g.

File: static/openapi/vwrs-spec.json (hunk 15 above)
Comment (joubert-macrometa) [reply static/openapi/vwrs-spec.json]: Fixed.

**Pull request merged**